executor/sort.go

```diff
@@ -119,85 +115,136 @@ func (e *SortExec) Open(ctx context.Context) error {
-	if err != nil {
-		return err
-	}
+func (e *SortExec) generatePartition() {
 	e.initPointers()
 	sort.Slice(e.rowPtrs, e.keyColumnsLess)
-	listInDisk, err := e.spillToDiskByRowPtr()
-	if err != nil {
-		return err
+	e.partitionList = append(e.partitionList, e.rowChunks)
+	e.partitionRowPtrs = append(e.partitionRowPtrs, e.rowPtrs)
+}
+
+type partitionPointer struct {
+	row         chunk.Row
+	partitionID int
+	consumed    int
+}
+
+type multiWayMerge struct {
+	lessRowFunction func(rowI chunk.Row, rowJ chunk.Row) bool
+	elements        []partitionPointer
+}
+
+func (h *multiWayMerge) Less(i, j int) bool {
+	rowI := h.elements[i].row
+	rowJ := h.elements[j].row
+	return h.lessRowFunction(rowI, rowJ)
+}
+
+func (h *multiWayMerge) Len() int {
+	return len(h.elements)
+}
+
+func (h *multiWayMerge) Push(x interface{}) {
+	// Should never be called.
+}
+
+func (h *multiWayMerge) Pop() interface{} {
+	h.elements = h.elements[:len(h.elements)-1]
+	return nil
+}
+
+func (h *multiWayMerge) Swap(i, j int) {
+	h.elements[i], h.elements[j] = h.elements[j], h.elements[i]
+}
+
+func (e *SortExec) externalSorting(req *chunk.Chunk) (err error) {
+	if e.heapSort == nil {
+		e.heapSort = &multiWayMerge{e.lessRow, make([]partitionPointer, 0, len(e.partitionList))}
+		for i := 0; i < len(e.partitionList); i++ {
+			row, err := e.partitionList[i].GetRow(e.partitionRowPtrs[i][0])
+			if err != nil {
+				return err
+			}
+			e.heapSort.elements = append(e.heapSort.elements, partitionPointer{row: row, partitionID: i, consumed: 0})
+		}
+		heap.Init(e.heapSort)
 	}
-	e.memTracker.Consume(-e.rowChunks.GetMemTracker().BytesConsumed())
-	e.rowChunks = nil
-	e.partitionList = append(e.partitionList, listInDisk)
-	e.partitionRowPtrs = append(e.partitionRowPtrs, e.initPointersForListInDisk(listInDisk))
-	return err
+
+	for !req.IsFull() && e.heapSort.Len() > 0 {
+		partitionPtr := e.heapSort.elements[0]
+		req.AppendRow(partitionPtr.row)
+		partitionPtr.consumed++
+		if partitionPtr.consumed >= len(e.partitionRowPtrs[partitionPtr.partitionID]) {
+			heap.Remove(e.heapSort, 0)
+			continue
+		}
+		partitionPtr.row, err = e.partitionList[partitionPtr.partitionID].GetRow(e.partitionRowPtrs[partitionPtr.partitionID][partitionPtr.consumed])
+		if err != nil {
+			return err
+		}
+		e.heapSort.elements[0] = partitionPtr
+		heap.Fix(e.heapSort, 0)
+	}
+	return nil
 }
 
 func (e *SortExec) fetchRowChunks(ctx context.Context) error {
 	fields := retTypes(e)
-	e.rowChunks = chunk.NewList(fields, e.initCap, e.maxChunkSize)
+	e.rowChunks = chunk.NewRowContainer(fields, e.maxChunkSize)
 	e.rowChunks.GetMemTracker().AttachTo(e.memTracker)
 	e.rowChunks.GetMemTracker().SetLabel(rowChunksLabel)
+	var onExceededCallback func(chk *chunk.Chunk)
+	if config.GetGlobalConfig().OOMUseTmpStorage {
+		e.spillAction = e.rowChunks.ActionSpill()
+		e.ctx.GetSessionVars().StmtCtx.MemTracker.FallbackOldAndSetNewAction(e.spillAction)
+		onExceededCallback = func(chk *chunk.Chunk) {
```

Comment:
Just inputting the last Chunk `chk` seems useless, how about replacing it with `*RowContainer`?

Reply:
Good idea. I will address.

Reply:
Addressed. PTAL @qw4990